src/integration/java/io/pinecone/helpers/IndexManagerSingleton.java

```diff
@@ -0,0 +1,211 @@
+package io.pinecone.helpers;
+
+import io.pinecone.clients.Index;
+import io.pinecone.clients.Pinecone;
+import io.pinecone.configs.PineconeConfig;
+import io.pinecone.exceptions.PineconeException;
+import io.pinecone.proto.DescribeIndexStatsResponse;
+import org.openapitools.client.model.*;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Arrays;
+import java.util.List;
+
+import static io.pinecone.helpers.BuildUpsertRequest.buildRequiredUpsertRequestByDimension;
+import static io.pinecone.helpers.IndexManager.*;
+import static org.junit.jupiter.api.Assertions.fail;
+
+public class IndexManagerSingleton {
+    private static final Logger logger = LoggerFactory.getLogger(IndexManager.class);
+    private static IndexManagerSingleton instance;
+    private static final String apiKey = System.getenv("PINECONE_API_KEY");
+    private final int dimension = System.getenv("DIMENSION") == null
+            ? 4
+            : Integer.parseInt(System.getenv("DIMENSION"));
+    private final String environment = System.getenv("PINECONE_ENVIRONMENT") == null
+            ? "us-east4-gcp"
+            : System.getenv("PINECONE_ENVIRONMENT");
+    private static final IndexMetric metric = System.getenv("METRIC") == null
+            ? IndexMetric.COSINE
+            : IndexMetric.valueOf(System.getenv("METRIC"));
+    private static final PineconeConfig pineconeConfig = new PineconeConfig(apiKey);
+    private static Pinecone pineconeClient;
+    private String podIndexName;
+    private IndexModel podIndexModel;
+    private String serverlessIndexName;
+    private IndexModel serverlessIndexModel;
+    private String collectionName;
+    private CollectionModel collectionModel;
+    private final List<String> podIndexVectorIds = Arrays.asList("v1", "v2", "v3");
+
+    private IndexManagerSingleton() {
+        pineconeClient = new Pinecone.Builder(apiKey).build();
+    }
+
+    public static IndexManagerSingleton getInstance() {
+        if (instance == null) {
+            instance = new IndexManagerSingleton();
+        }
+        return instance;
+    }
+
+    public String getPodIndexName() throws InterruptedException, PineconeException {
+        if (podIndexName == null) {
+            podIndexName = createPodIndex();
+        }
+        return podIndexName;
+    }
+
+    public String getServerlessIndexName() throws InterruptedException, PineconeException {
+        if (serverlessIndexName == null) {
+            serverlessIndexName = createServerlessIndex();
+        }
+        return serverlessIndexName;
+    }
+
+    public String getCollectionName() throws InterruptedException {
+        if (collectionName == null) {
+            collectionName = createCollection();
+        }
+        return collectionName;
```

Comment:
I would consider moving all of these existence conditionals to within the `createCollection` method, or `createServerlessIndex`, etc.

For example, if it was instead called `createOrGetServerlessIndex` you could use it like this:

```java
    public String getServerlessIndexName() throws InterruptedException, PineconeException {
        return createOrGetServerlessIndex();
    }

    private String createOrGetServerlessIndex() throws InterruptedException, PineconeException {
       // -----> null checks encapcsulated here
       if (serverlessIndexName == null) {
            return serverlessIndexName;
        }

        String indexName = RandomStringBuilder.build("serverless-index", 8);

        // Create index
        ServerlessSpec serverlessSpec = new ServerlessSpec().cloud(ServerlessSpec.CloudEnum.AWS).region("us-west-2");
        CreateIndexRequestSpec spec = new CreateIndexRequestSpec().serverless(serverlessSpec);
        CreateIndexRequest createIndexRequest = new CreateIndexRequest().name(indexName).dimension(dimension).metric(metric).spec(spec);
        serverlessIndexModel = pineconeClient.createIndex(createIndexRequest);

        // -------> have to save it here now
        this.serverlessIndexName = indexName;
        return indexName;
    }
```